File: jboot/modules/jboot-openapi/src/main/java/com/jboot/openapi/OpenApi.java
package com.jboot.openapi;

import com.jboot.openapi.api.*;
import com.jboot.openapi.core.*;

/**
 * OpenAPI/Swagger documentation generation utilities.
 * 
 * <p>
 * This module enables automatic API documentation generation from Java
 * handlers.
 * 
 * <h2>Example Usage:</h2>
 * 
 * <pre>{@code
 * var spec = OpenApiSpec.builder()
 *         .title("My API")
 *         .version("1.0.0")
 *         .description("API for managing users")
 *         .build();
 * 
 * spec.path("/users")
 *         .get(Operation.builder()
 *                 .summary("List all users")
 *                 .tag("users")
 *                 .response(200, Schema.arrayOf(User.class))
 *                 .build())
 *         .post(Operation.builder()
 *                 .summary("Create a user")
 *                 .tag("users")
 *                 .requestBody(Schema.of(CreateUserRequest.class))
 *                 .response(201, Schema.of(User.class))
 *                 .build());
 * 
 * // Generate JSON
 * String json = spec.toJson();
 * 
 * // Generate YAML
 * String yaml = spec.toYaml();
 * }</pre>
 */
public final class OpenApi {

    private OpenApi() {
    }

    /**
     * Creates a new OpenAPI specification builder.
     */
    public static OpenApiSpec.Builder spec() {
        return OpenApiSpec.builder();
    }

    /**
     * Creates a schema from a Java class using reflection.
     */
    public static Schema schemaOf(Class<?> clazz) {
        return SchemaGenerator.generate(clazz);
    }

    /**
     * Creates an array schema.
     */
    public static Schema arrayOf(Class<?> itemClass) {
        return Schema.builder()
                .type("array")
                .items(schemaOf(itemClass))
                .build();
    }

    /**
     * Creates a map schema.
     */
    public static Schema mapOf(Class<?> valueClass) {
        return Schema.builder()
                .type("object")
                .additionalProperties(schemaOf(valueClass))
                .build();
    }
}

File: goboot/openapi/openapi.go
// Package openapi provides OpenAPI/Swagger documentation generation utilities.
//
// This module enables automatic API documentation from Go handlers.
//
// Example:
//
//	spec := openapi.NewSpec("My API", "1.0.0")
//	spec.AddPath("/users", openapi.Path{
//	    Get: &openapi.Operation{
//	        Summary: "List users",
//	        Responses: openapi.Responses{
//	            "200": {Description: "Success", Schema: openapi.ArrayOf(User{})},
//	        },
//	    },
//	})
//
//	// Generate OpenAPI JSON
//	json := spec.ToJSON()
package openapi

import "dev.engineeringlabs/goboot/openapi/api"

// Re-export API types
type (
	// Spec represents an OpenAPI specification.
	Spec = api.Spec
	// Info contains API metadata.
	Info = api.Info
	// Path represents an API path.
	Path = api.Path
	// Operation represents an HTTP operation.
	Operation = api.Operation
	// Parameter represents an operation parameter.
	Parameter = api.Parameter
	// Response represents an operation response.
	Response = api.Response
	// Schema represents a JSON schema.
	Schema = api.Schema
	// Tag represents an API tag.
	Tag = api.Tag
	// Contact represents contact information.
	Contact = api.Contact
	// License represents license information.
	License = api.License
	// Server represents a server URL.
	Server = api.Server
	// SecurityScheme represents a security scheme.
	SecurityScheme = api.SecurityScheme
)

// Responses is a map of status codes to responses.
type Responses = map[string]Response

// NewSpec creates a new OpenAPI specification.
func NewSpec(title, version string) *Spec {
	return &Spec{
		OpenAPI: "3.0.3",
		Info: Info{
			Title:   title,
			Version: version,
		},
		Paths: make(map[string]Path),
	}
}

// ArrayOf creates an array schema for the given type.
func ArrayOf(item interface{}) Schema {
	return Schema{
		Type:  "array",
		Items: SchemaFrom(item),
	}
}

// SchemaFrom creates a schema from a Go struct.
func SchemaFrom(v interface{}) *Schema {
	// Implementation would use reflection to build schema
	return &Schema{Type: "object"}
}
